### ISSUE TITLE
DHT Bootstrap issue

### ISSUE THREAD

**Issue by jbenet** [jbenet/go-ipfs#572]
Currently, DHT bootstrap only runs when whole core bootstrap runs. Turns out this is not correct, because it conflates IpfsCore's bootstrap requirements with those of the dht. Namely:

- IPFS core bootstrap requirement: that we're connected to at least N nodes, where N is a tunable parameter related to the size of the number of bootstrap peers we use.
- DHT bootstrap requirement: that we have a well-formed routing table (i.e. we have a representative samle of peers in our table). 

What this yields is that nodes -- upon starting up -- will connect to a bunch of peers and run dht bootstrap. dht routing table will be well formed. As time passes, those connections may break momentarily. the core bootstrap routine doesnt do anything about it, because the node may be connected to more than N bootstrap nodes, so it skips bootstrap altogether. Meanwhile, the dht's routing table worsens, but the dht doesn't do anything about it. 

Note that this is not technically the core's problem to fix-- it's requirements on being connected to N bootstrap peers are correct. And, as far as it's concerned, it called `dht.Bootstrap()` at least once. Now, we _may_ lift the requirement that ipfs Bootstrap should include good Routing system bootstrap conditions, or not:

2 proposed solutions:

1. `(*IpfsCore).Bootstrap()` periodic routine _always_ calls `dht.Bootstrap()` (never skips).
2. `IpfsDHT` uses its own async worker to periodically call run its own bootstrapping logic. (spawned by either a call to `Bootstrap`, or a call to `NewDHT`. 

(1) is simpler on the `DHT`. (2) is simpler on the `Core`.

Example of 1:
 ```go
 func (core *IpfsCore) Bootstrap() {
 	...
 	// _always_ call Bootstrap. routing can decide
 	// what "being well bootstrapped" means.
 	core.Routing.Bootstrap()
 	...
 }
```

Example of 2:

```go
dht := NewDHT(...)
dht.Bootstrap()
// dht is ready to be used, and will continue usable.
```

Feedback on the approaches appreciated, i'm atm leaning towards (2). cc @briantigerchow @whyrusleeping

**Comment by whyrusleeping** [comment]
I think the DHT running its own async worker that periodically bootstraps to maintain its routing table is going to be the best option

**Comment by briantigerchow** [comment]
If the DHT knows best, then it may not make sense to get callers (IpfsNode et al) involved.

If the dht _exposes_ bootstrapping functionality in its interface, `dht.Bootstrap()` should only perform the operation once and should either a) block during the operation or b) return a future. This way callers retain control over the object.

Periodic bootstrapping might work best as a constructor option.

**Comment by whyrusleeping** [comment]
I think the call to start the boostrapping routine should be separate, we might have a scenario where we wish to construct a DHT, but not start the bootstrapping, so:

```
ipfsdht := dht.NewDht(...)
go ipfsdht.Bootstrap()
```

**Comment by briantigerchow** [comment]
this is why I mention it as a constructor option. The lifecycle of the bootstrap routine is tied to the lifecycle of the dht so it's best to let the dht manage that internally.

**Comment by whyrusleeping** [comment]
ooooh, i see what you mean. :+1:

**Comment by jbenet** [comment]
> this is why I mention it as a constructor option. 

As it stands now, this wont work correctly: we instantiate the DHT well before bootstrap finishes. TRTTD is for the network passed in to have some notion of being "online or offline", and the dht bootstrap to wait on that. However that requires lots of changes to how the interfaces work right now, and IMO not worth it to fix this one fix atm. (this can be the way going fwd though, but should be designed carefully).

I'll opt for something in the middle: 

```go
// the bootstrap function returns a process
func (dht *IpfsDHT) Boostrap() goprocess.Process { ... }

// if desired, the caller can store that procsess
bsproc := dht.Bootstrap()

// and stop it when they wish
bsproc.Close()
```

**Comment by jbenet** [comment]
the caller can treat it as a simple `io.Closer`:

```go
var bsproc io.Closer = dht.Bootstrap()
bsproc.Close()
```

but returning a goprocess so the caller can also wire the signals:

```go
bsproc := dht.Bootstrap()
anotherProc.AddChild(bsproc)
anotherProc.Close() // closes bsproc too
```

**Comment by briantigerchow** [comment]
(The goprocess should be able to manage closers. It simplifies the interface for users since the only thing the process could possibly do is close the argument. This occurred to be yesterday with the context closer.)

https://github.com/jbenet/go-ctxgroup/commit/f6f9601038d519c1c72e0f3bf4b55b1aa5120997

```Go
var bsproc io.Closer = dht.Bootstrap()
anotherProc.Foo(bsproc) // closes closer 
```

**Comment by jbenet** [comment]
yep, i saw: https://github.com/jbenet/go-ctxgroup/pull/1#issuecomment-70187058

And, that's not exactly correct, there is value in knowing it's a process and not just a closer: https://github.com/jbenet/goprocess/blob/master/impl-mutex.go#L138

goprocess can do this by attempting a cast, and wrapping it in a process otherwise that captures the error properly. One of the important features of goprocess is that it calls Close with **at-most-once** semantics, and thus never loses the returned `error`. note many closers are written such that subsequent calls to close return nil. This works in spite of a process receiving multiple signals to close-- all callers receive the same `error`.

**Comment by jbenet** [comment]
Addressed in #583
